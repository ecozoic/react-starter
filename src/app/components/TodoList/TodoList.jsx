import React from 'react';
import PropTypes from 'prop-types';

import { todoShape } from '../../models';
import TodoComponent from '../Todo/index';

import styles from './TodoList.scss';

const TodoList = ({ todos, onTodoClick }) => (
  <ul className={styles.todoList}>
    {todos.map(todo => (
      <TodoComponent
        key={todo.id}
        onClick={() => onTodoClick(todo.id)}
        completed={todo.completed}
        text={todo.text}
      />
    ))}
  </ul>
);

TodoList.propTypes = {
  todos: PropTypes.arrayOf(todoShape).isRequired,
  onTodoClick: PropTypes.func.isRequired,
};

export default TodoList;
