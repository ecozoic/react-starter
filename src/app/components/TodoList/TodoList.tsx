import React from 'react';
import PropTypes from 'prop-types';
import CircularProgress from 'material-ui/CircularProgress';

import { todoShape, Todo } from '../../models';
import TodoComponent from '../Todo';

const styles = require('./TodoList.scss');

export interface TodoListProps {
  todos: Todo[];
  onTodoClick: (id: number) => void;
  loading?: boolean;
}

const TodoList: React.SFC<TodoListProps> = ({ todos, onTodoClick, loading }) => {
  return (
    <div>
      {loading ? (
        <div className={styles.progressContainer}>
          <CircularProgress />
        </div>
      ) : (
        <ul className={styles.todoList}>
          {todos.map(todo => (
            <TodoComponent
              key={todo.id}
              onClick={() => onTodoClick(todo.id)}
              completed={todo.completed}
              text={todo.text}
            />
          ))}
        </ul>
      )}
    </div>
  );
};

TodoList.propTypes = {
  todos: PropTypes.arrayOf(todoShape).isRequired,
  onTodoClick: PropTypes.func.isRequired,
  loading: PropTypes.bool,
};

TodoList.defaultProps = {
  loading: false,
};

export default TodoList;
