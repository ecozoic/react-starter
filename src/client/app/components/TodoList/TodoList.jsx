/* @flow */
import React, { PropTypes } from 'react';
import { StyleSheet, css } from 'aphrodite';

import Todo from '../Todo';

const styles = StyleSheet.create({
  todoList: {
    cursor: 'pointer',
  },
});

function TodoList({ todos, onTodoClick }) {
  return (
    <ul className={css(styles.todoList)}>
      {todos.map(todo =>
        <Todo
          key={todo.id}
          {...todo}
          onClick={() => onTodoClick(todo.id)}
        />,
      )}
    </ul>
  );
}

TodoList.propTypes = {
  todos: PropTypes.arrayOf(
    PropTypes.shape({
      id: PropTypes.number.isRequired,
      completed: PropTypes.bool.isRequired,
      text: PropTypes.string.isRequired,
    }).isRequired,
  ).isRequired,
  onTodoClick: PropTypes.func.isRequired,
};

export default TodoList;
