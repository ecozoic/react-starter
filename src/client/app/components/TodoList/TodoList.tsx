import * as React from 'react';
import { Component } from 'react';
import * as PropTypes from 'prop-types';

import { todoShape, Todo } from '../../models';
import TodoComponent from '../Todo';

const styles = require('./TodoList.scss');

export interface TodoListProps {
  todos: Todo[];
  onTodoClick: (id: number) => void;
}

class TodoList extends Component<TodoListProps, undefined> {
  static propTypes = {
    todos: PropTypes.arrayOf(todoShape).isRequired,
    onTodoClick: PropTypes.func.isRequired,
  };

  render() {
    const { todos, onTodoClick } = this.props;

    return (
      <ul className={styles.todoList}>
        {todos.map((todo) => (
          <TodoComponent
            key={todo.id}
            onClick={() => onTodoClick(todo.id)}
            completed={todo.completed}
            text={todo.text}
          />
        ))}
      </ul>
    );
  }
}

export default TodoList;
