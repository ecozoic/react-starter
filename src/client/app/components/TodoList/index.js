/* @flow */
export { default } from './TodoList';
