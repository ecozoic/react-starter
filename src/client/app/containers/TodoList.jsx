/* @flow */
import { connect } from 'react-redux';

import { toggleTodo } from '../actions';
import TodoList from '../components/TodoList/TodoList';

const mapStateToProps = (state) => {
  return {
    todos: state.get('todos').toJS(),
  };
};

const mapDispatchToProps = (dispatch) => {
  return {
    onTodoClick: (id) => {
      dispatch(toggleTodo(id));
    },
  };
};

const TodoListContainer = connect(
  mapStateToProps,
  mapDispatchToProps,
)(TodoList);

export default TodoListContainer;
