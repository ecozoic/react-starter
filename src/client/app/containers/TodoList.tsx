import { connect } from 'react-redux';

import { toggleTodo } from '../actions';
import { getTodos } from '../selectors';
import TodoList from '../components/TodoList';

const mapStateToProps = (state: any) => ({
  todos: getTodos(state),
});

const mapDispatchToProps = (dispatch: Function) => ({
  onTodoClick: (id: number) => dispatch(toggleTodo(id)),
});

const TodoListContainer = connect(
  mapStateToProps,
  mapDispatchToProps,
)(TodoList);

export default TodoListContainer;
