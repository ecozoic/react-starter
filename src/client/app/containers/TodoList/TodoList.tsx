import { connect } from 'react-redux';
import { Dispatch } from 'redux';

import { toggleTodo } from '../../actions';
import { getTodos } from '../../selectors';
import { State } from '../../reducers';
import TodoList from '../../components/TodoList';

const mapStateToProps = (state: State) => ({
  todos: getTodos(state),
});

const mapDispatchToProps = (dispatch: Dispatch<State>) => ({
  onTodoClick: (id: number) => {
    dispatch(toggleTodo(id));
  },
});

const TodoListContainer = connect(
  mapStateToProps,
  mapDispatchToProps,
)(TodoList);

export default TodoListContainer;
