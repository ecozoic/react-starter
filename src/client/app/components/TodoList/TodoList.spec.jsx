/* @flow */
import React from 'react';
import { mount, shallow } from 'enzyme';

import TodoList from './TodoList';

describe('TodoList', () => {
  it('renders without crashing', () => {
    const todoList = mount(
      <TodoList
        todos={[]}
        onTodoClick={function noop() { }}
      />,
    );

    expect(todoList).toMatchSnapshot();
  });

  describe('render', () => {
    it('displays an unordered list', () => {
      const todoList = shallow(
        <TodoList
          todos={[]}
          onTodoClick={function noop() { }}
        />,
      );

      expect(todoList.find('ul').length).toBe(1);
    });

    it('renders a todo for each todo in props', () => {
      const todoList = shallow(
        <TodoList
          todos={[
            { id: 1, text: 'Todo 1', completed: false },
            { id: 2, text: 'Todo 2', completed: true },
            { id: 3, text: 'Todo 3', completed: false },
          ]}
          onTodoClick={function noop() { }}
        />,
      );

      expect(todoList.find('Todo').length).toBe(3);
    });

    it('renders todo with proper props', () => {
      const todoList = shallow(
        <TodoList
          todos={[
            { id: 1, text: 'Todo 1', completed: false },
          ]}
          onTodoClick={function noop() { }}
        />,
      );

      const todo = todoList.find('Todo');

      expect(todo.props().completed).toBe(false);
      expect(todo.props().text).toBe('Todo 1');
    });

    it('sets the todo id as the key', () => {
      const todoList = shallow(
        <TodoList
          todos={[
            { id: 1, text: 'Todo 1', completed: false },
          ]}
          onTodoClick={function noop() { }}
        />,
      );

      expect(todoList.find('Todo').key()).toEqual('.$1');
    });

    it('calls onTodoClick with id of clicked todo', () => {
      const mockFn = jest.fn();

      const todoList = shallow(
        <TodoList
          todos={[
            { id: 1, text: 'Todo 1', completed: false },
          ]}
          onTodoClick={mockFn}
        />,
      );

      todoList.find('Todo').simulate('click');

      expect(mockFn).toHaveBeenCalledTimes(1);
      expect(mockFn.mock.calls[0][0]).toBe(1);
    });
  });
});
