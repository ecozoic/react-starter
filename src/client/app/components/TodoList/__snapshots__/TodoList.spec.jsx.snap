// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoList renders without crashing 1`] = `
<TodoList
  onTodoClick={[Function]}
  todos={Array []}
>
  <ul
    className="TodoList"
  />
</TodoList>
`;
